plugins/goodix-fp/fu-goodixfp-device.c

```diff
@@ -0,0 +1,362 @@
+			     GxfpCmdResp *presponse,
+			     gboolean data_reply,
+			     GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE(device));
+
+	if (!goodixfp_device_cmd_send (usb_device, cmd0, cmd1, pkg_eop, request, error))
+		return FALSE;
+	return goodixfp_device_cmd_recv (usb_device, presponse, data_reply, error);
+}
+
+static gchar *
+fu_goodixfp_device_get_version (FuGoodixFpDevice *self, GError **error)
+{
+	GxfpCmdResp reponse = {0,};
+	guint8 dummy = 0;
+	gchar ver[9] = {0};
+
+	g_autoptr(GByteArray) request = g_byte_array_new ();
+	fu_byte_array_append_uint8 (request, dummy);
+	if (!fu_goodixfp_device_cmd_xfer (self, GX_CMD_VERSION, GX_CMD1_DEFAULT,
+					  0,
+					  request,
+					  &reponse,
+					  TRUE,
+					  error))
+		return NULL;
+	memcpy (ver, reponse.version_info.fwversion, 8);
+	return g_strdup (ver);
+}
+
+static gboolean
+fu_goodixfp_device_update_init(FuGoodixFpDevice *self, GError **error)
+{
+	GxfpCmdResp reponse = { 0,};
+	g_autoptr(GByteArray) request = g_byte_array_new ();
+	/* update initial */
+	if (!fu_goodixfp_device_cmd_xfer (self, GX_CMD_UPGRADE, GX_CMD_UPGRADE_INIT,
+					  0,
+					  request,
+					  &reponse,
+					  TRUE,
+					  error))
+		return FALSE;
+	return (reponse.result == 0);
+}
+
+static gboolean
+fu_goodixfp_device_attach (FuDevice *device, GError **error)
+{
+	FuGoodixFpDevice *self = FU_GOODIXFP_DEVICE(device);
+	g_autoptr(GByteArray) request = g_byte_array_new ();
+	GxfpCmdResp reponse = {0,};
+
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	/* reset device */
+	if (!fu_goodixfp_device_cmd_xfer (self, GX_CMD_RESET, 0x03,
+					  0,
+					  request,
+					  &reponse,
+					  FALSE,
+					  error))
+		return FALSE;
+	if (reponse.result != 0)
+		return FALSE;
+
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_goodixfp_device_open (FuUsbDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (device);
+
+	if (!g_usb_device_claim_interface (usb_device, GX_USB_INTERFACE,
+					   G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					   error))
+		return FALSE;
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_goodixfp_device_setup (FuDevice *device, GError **error)
+{
+	FuGoodixFpDevice *self = FU_GOODIXFP_DEVICE(device);
+	g_autofree gchar *version = NULL;
+	version = fu_goodixfp_device_get_version (self, error);
+	if (!version) {
+		g_prefix_error (error, "failed to get firmware version: ");
+		return FALSE;
+	}
+	g_debug ("obtained fwver using API '%s'", version);
+	fu_device_set_version (device, version);
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_goodixfp_device_write_firmware (FuDevice *device,
+				   FuFirmware *firmware,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	FuGoodixFpDevice *self = FU_GOODIXFP_DEVICE(device);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+	g_autoptr(GError) error_local = NULL;
```

Comment:
we tend to put the `g_auto` stuff as the last section of the C99 variable block, i.e. after GxfpCmdResp.